### PR TITLE
Pass hosts/ports to computed processes from orchestrators, enabling multi-process instances.

### LINKED ISSUE #12053: Properly compute peer hosts/ports from env vars in Kubernetes-orchestrated `computed`
### What version of Materialize are you using?

main

### How did you install Materialize?

Materialize Cloud

### What was the issue?

See discussion on Slack: https://materializeinc.slack.com/archives/C02KE368L8P/p1650940659088839

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by umanwizard**
### Motivation

Fixes #12053

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Pass correct command-line arguments for multi-process compute clusters. (4xlarge and above)

### CODE REVIEW

File: src/compute/src/bin/computed.rs [hunk 1]
```diff
@@ -122,6 +116,31 @@ fn create_communication_config(args: &Args) -> Result<timely::CommunicationConfi
...
+                        .rsplit_once('-')
+                        .and_then(|(_, idx)| idx.parse::<usize>().ok())
+                })
+            })
+            .ok_or_else(|| {
+                anyhow::anyhow!(
+                    "Process index not provided, and failed to deduce it from hostname."
+                )
+            })?
+    };
```
Comment (benesch) [inline src/compute/src/bin/computed.rs:143]: Logic looks good, but I was hoping to keep it in src/compute/ci/Dockerfile, where a shell entrypoint could parse `HOSTNAME` and export it as `COMPUTED_PROCESS`. That isolates the weirdness from the process orchestrator.

File: src/orchestrator-process/src/lib.rs [hunk 2]
```diff
@@ -135,7 +136,11 @@ impl NamespacedOrchestrator for NamespacedProcessOrchestrator {
                     .ok_or_else(|| anyhow!("port exhaustion"))?;
                 ports.insert(port.name.clone(), p);
             }
-            let args = args(&ports);
+            hosts_ports.push(("localhost".to_string(), ports))
+        }
+        for i in 0..(scale_in.get()) {
+            let ports = &hosts_ports[i].1;
+            let args = args(&hosts_ports, ports, Some(i));
```
Comment (benesch) [inline src/orchestrator-process/src/lib.rs:143]: This looks exactly right, but unfortunately it is a good bit harder with @jkosh44's latest changes—you'll see when you rebase.

File: src/orchestrator/src/lib.rs [hunk 3]
```diff
@@ -77,10 +77,23 @@ pub struct ServiceConfig<'a> {
...
+    /// configures the arguments for all processes at once.
     #[derivative(Debug = "ignore")]
-    pub args: &'a (dyn Fn(&HashMap<String, i32>) -> Vec<String> + Send + Sync),
+    pub args: &'a (dyn Fn(
+        &[(String, HashMap<String, i32>)],
+        &HashMap<String, i32>,
+        Option<usize>,
+    ) -> Vec<String>
+             + Send
+             + Sync),
```
Comment (benesch) [inline src/orchestrator/src/lib.rs:96]: 👍🏽

**Review by benesch — APPROVED** [review]
Nice! This is exactly what I had in mind.

File: src/orchestrator-process/src/lib.rs (hunk 2 above)
Comment (umanwizard) [reply src/orchestrator-process/src/lib.rs:143]: Conceptually it's simple, no? You can still break apart the loop and precompute all the mappings, you just need to keep track of the ones you read the  ports out of a file for so you can avoid passing the corresponding machines to `supervise` in the second loop.

Comment (benesch) [reply src/orchestrator-process/src/lib.rs:143]: Yep, definitely. It's just mechanically more difficult!

**Comment by umanwizard** [comment]
@jkosh44 Would appreciate you taking a look too since I mucked with the portfile code.

File: src/orchestrator-process/src/lib.rs [hunk 4]
```diff
@@ -182,24 +191,36 @@ impl NamespacedOrchestrator for NamespacedProcessOrchestrator {
...
-                    Arc::clone(&self.port_allocator),
-                    ports,
-                    self.suppress_output,
-                    port_metadata_file_location,
-                ),
-            ))));
+        }
+
+        // Now create all the processes that weren't detected as being still alive
+        // let hosts_ports = std::iter::from_fn(|| ("localhost".to_string(), ))
```
Comment (umanwizard) [inline src/orchestrator-process/src/lib.rs:198]:
```suggestion
        // Now create all the processes that weren't detected as being still alive
```

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/orchestrator-process/src/lib.rs [hunk 5]
```diff
@@ -182,24 +191,35 @@ impl NamespacedOrchestrator for NamespacedProcessOrchestrator {
...
+                ));
+
+                handles.push(AbortOnDrop(Box::new(mz_ore::task::spawn(
+                    || format!("service-supervisor: {full_id}"),
+                    supervise(
+                        full_id.clone(),
+                        path.clone(),
+                        args.clone(),
+                        Arc::clone(&self.port_allocator),
+                        std::mem::take(&mut processes[i]),
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/orchestrator-process/src/lib.rs]: yes lol I was literally trying to figure out why cargo test was failing when you wrote that

**Pull request merged**